### ISSUE TITLE
Support for direct access to files in Win95/98

### ISSUE THREAD

**Issue by AP2020C** [joncampbell123/dosbox-x#5357]
### Is your feature request related to a problem? Please describe.

What I have always appreciated about DosBox is the ability to use an existing real windows folder as C, consequently, having all the files accessible from the host system.
When I try to instal windows 95/98 I am forced to create a static size img file.

I would like to:
* variable size disks based on content, and with the ability to create snapshots (like classic virtual machines).
* that DosBoxX would natively take care of transforming disk calls to the format windows likes even using a folder and not an image
* to have ready-made, in the various languages, installations of the main operating systems, preinstalled, on a par with Dos

### What you want

* variable size disks based on content, and with the ability to create snapshots (like classic virtual machines).
* that DosBoxX would natively take care of transforming disk calls to the format windows likes even using a folder and not an image
* to have ready-made, in the various languages, installations of the main operating systems, preinstalled, on a par with Dos

### Have you checked that no similar feature request(s) exist?

- [X] I have searched and didn't find any similar feature request.

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

**Comment by DamjanJovanovic** [comment]
There are some interesting ways to implement this feature.

Windows 9x supported user-installable custom filesystems, so a VxD device driver for Windows 9x could be developed that emulates a disk and filesystem containing the files from the host. I wanted to do this for VirtualBox's "guest extensions" too. This would allow maximum performance and best integration with the host, but the development complexity is extreme: not only does such a driver run in kernel mode, and need to integrate with the awful Win9x kernel APIs, but it needs to use Win9x's caching for good performance, deal with short and long filenames, emulate a virtual disk to trick Win9x into loading this filesystem for it during its automatic mounting process (because there is no other way to register a filesystem), and is generally a lot of work. Also, this custom filesystem wouldn't help you install Win9x, as the installer is a completely different process, running in DOS IIRC, so at best you'd have to install to a disk image that you continue to boot Win9x from, but then maybe use the custom filesystem to access additional files on the host.

But there is another brilliant idea I heard about. The FAT16/FAT32 filesystem is very simple and very well understood, and widely supported by many operating systems. There is a module in qemu, vvfat (virtual VFAT) (https://en.wikibooks.org/wiki/QEMU/Devices/Storage) which "shadows a local directory". On startup, before booting the guest OS, this module scans all the files in the given directory, and builds a virtual FAT filesystem containing pointers to these files and offsets within them, which it presents to the guest OS as a "disk". When the guest OS tries to read a block from this "disk", this module can figure out which file/offset is at that block, and go and do a read from the host file, successfully fooling the guest OS that it read from its "disk" as intended. Later versions also implemented writing files. Thus, by emulating a FAT16/FAT32 "disk" from underlying host files, like the qemu vvfat module does, DOS and Win9x would easily access host files using their preferred native filesystem driver. This does not require installation of any other drivers into DOS or Win9x. It might even be able to boot Win9x. But startup would be slower as every shared host file has to be scanned (we need its name and size ahead of time, to be able to place it somewhere on the virtual disk, so when its blocks are requested we know which file and offset to read from), slightly higher memory usage to store all filenames, inability to synchronize certain operations with host file (for example, byte-range locking is invisible to the host), also any new files the host creates in the shared directory are invisible to the guest until reboot (they have to be - the guest caches directory contents and avoid reading disk for performance, and would only rescan a directory from disk (and discover the new files the host wrote) when the cache expires), and the shared files cannot be edited or moved or deleted by the host while the guest is running. Qemu's source code for vvfat is at: https://github.com/qemu/qemu/blob/master/block/vvfat.c

**Comment by DamjanJovanovic** [comment]
> Win9x actually can still use DOS API to access disks. It was designed as 'fallback' in case of no native drivers and supposed to be slower.

Yes, Raymond Chen explained how that worked: https://devblogs.microsoft.com/oldnewthing/20071224-00/?p=24063

But we care about filesystems, not disks. Does Win9x use MS-DOS filesystems? The book "Undocumented DOS" by Andrew Schulman et al, 1990, chapter 4, the "The DOS file system and network redirector", reverse engineered an undocumented API which allows creating custom filesystems for MS-DOS, and which was used to create the MSCDEX filesystem.

Now Win9x certainly had VxD drivers for both FAT and ISO9660 filesystems, and thus could avoid needing DOS's own FAT and MSCDEX code, but what if other custom DOS filesystem drivers were installed? Did Win9x use them? Even if it did, it would have been a poor experience, as serializing I/O through DOS is a slow, synchronous operation, while direct 32 bit I/O to the IDE controller, which the Win9x VxD filesystems can do, and which the Qemu vvfat module's approach can also facilitate, is fast and asynchronous.

> Long time ago there was commercial Win9x emulator called Win4Lin: https://en.wikipedia.org/wiki/Win4Lin It was able to boot DOS or Win9x from Linux directory.
> 
> But to make it work with DosBox it'll be required to implement all extra Windows-specific DOS enhancements from MS-DOS 7.x...

Impressive. Any idea how they did it?

**Comment by dionorgua** [comment]
@DamjanJovanovic I think it's exactly same thing like DosBox: implement all I/O functions from INT21. I'm sure that Win98 still can use it (in case of no 'native' drivers). But 'faster' 32 bit way was to use native drivers (that provides access to block devices) plus filesystem drivers.

But I can be wrong. Win4lin was designed to run only certain versions of Win9x systems.. I think they were also patching certain binaries during 'install'. Most likely to run Ring0 code in userspace.. But it could happens that also for filesystem access. But I'm sure that booting Win4lin was much faster that booting Win9x natively on hardware.

See also this video: https://www.youtube.com/watch?v=TTA3b0i9n_Q

The problem is that DosBox DOS implementation is not good enough to run Windows: https://github.com/joncampbell123/dosbox-x/issues/1217

PS. Thanks for links. Will take a look

**Comment by DamjanJovanovic** [comment]
My tests with dosbox-x 2025.02.01 show Windows 95 can already see host folder mounts ("MOUNT D /path") as drives, and browse them and view their files, but filesystem changes it makes there strangely do not appear on the host and disappear on reboot.

In Control Panel, System, on the Performance tab, it shows "Drive D: using MS-DOS compatibility mode file system." (for all drives, not just D:).

It's quite impressive that it works, and means that even though neither dosbox nor dosbox-x implements that INT 0x2f "network director" API for 3rd party filesystems from the "Undocumented DOS" book I mentioned, Windows can still figure out a custom DOS filesystem is being used, and continues to use it through DOS calls.

But why do writes to host folders appear to succeed in the guest, creating files and directories, but then the host doesn't see them and they disappear on reboot?

**Comment by DamjanJovanovic** [comment]
> But why do writes to host folders appear to succeed in the guest, creating files and directories, but then the host doesn't see them and they disappear on reboot?

From reading the code, I think my localDrive is somehow wrapped by an Overlay_Drive.

**Comment by DamjanJovanovic** [comment]
> You cannot write to folder mounts from Win9x, this is a known limitation and mentioned on the wiki.

But qemu's vvfat module let you do it, so clearly it's possible. What stops dosbox-x from implementing it? Is the Overlay_Drive wrapper inserted as a safety precaution, since booting from the boot sector ignores DOS drive numbering? Or what happens and why?

**Comment by maron2000** [comment]
Why not do it yourself?
No one stops you from contributing to introducing a new feature.
